app/src/main/java/net/aiscope/gdd_app/ui/mask/customview/MaskLayer.kt

```diff
@@ -46,171 +35,225 @@ class MaskLayer(
+        currentPaintEraser = newDefaultPaintEraser(calculateCurrentStrokeWidth())
+        paintEraserPendingRecreation = false
     }
 
-    private fun pxToDp(px: Int) = (px * context.resources.displayMetrics.density).toInt()
+    private fun resolveCurrentPaint() {
+        currentPaint = if (isCurrentModeDraw()) currentPaintBrush else currentPaintEraser
+    }
 
-    fun onViewSizeChanged(width: Int, height: Int) {
-        viewDimensions = width to height
-        scaleBrushAndTextPaints()
+    fun drawMode() {
+        currentMode = MaskCustomView.Mode.Draw
     }
 
-    fun onScaleChanged() = scaleBrushAndTextPaints()
+    fun eraseMode() {
+        currentMode = MaskCustomView.Mode.Erase
+    }
 
     fun undo() = undoPendingPaths++
 
     fun redo() = undoPendingPaths--
 
-    fun undoAvailable() = (pathsPaintsAndStagesNames.size - undoPendingPaths) > 0
+    fun undoAvailable() = (pathsAndPaints.size - undoPendingPaths) > 0
 
     fun redoAvailable() = undoPendingPaths > 0
 
     fun getBitmap(): Bitmap {
-        val (width, height) = bitmapDimensions
-        return Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888).apply {
+        return Bitmap.createBitmap(
+            currentStateBitmap.width,
+            currentStateBitmap.height,
+            Bitmap.Config.ARGB_8888
+        ).apply {
             val canvas = Canvas(this)
             drawPaths(canvas, removeAlpha = true)
         }
     }
 
     fun drawStart(x: Float, y: Float) {
+        refreshPaints()
         currentPath = PointToPointPath(x, y)
     }
 
+    private fun refreshPaints() {
+        if (paintBrushPendingRecreation) refreshPaintBrush()
+        if (paintEraserPendingRecreation) refreshPaintEraser()
+        resolveCurrentPaint()
+    }
+
     fun drawMove(x: Float, y: Float) {
-        val (latestX, latestY) = currentPath.latestPoint
+        val (latestX, latestY) = currentPath!!.latestPoint
         val dX = abs(x - latestX)
         val dY = abs(y - latestY)
-        val touchTolerance = ViewConfiguration.get(context).scaledTouchSlop / currentScale()
+        val touchTolerance = ViewConfiguration.get(context).scaledTouchSlop / currentScale
         if (dX >= touchTolerance || dY >= touchTolerance) {
-            currentPath.quadTo(x, y)
-            if (pathsPaintsAndStagesNames.isEmpty() || currentPath != pathsPaintsAndStagesNames.last().path) {
-                addPath(currentPath)
-            }
+            currentPath!!.quadTo(x, y)
         }
     }
 
-    fun onSaveInstanceState() =
-        MaskCustomViewBaseState(pathsPaintsAndStagesNames, undoPendingPaths, brushDiseaseStage)
-
-    fun onRestoreInstanceState(savedState: Parcelable) {
-        if (savedState is MaskCustomViewBaseState) {
-            undoPendingPaths = savedState.undoPendingPaths
-            pathsPaintsAndStagesNames.addAll(savedState.reassemblePathsPaintsAndStagesNames())
-            brushDiseaseStage = savedState.currentBrushDiseaseStage
+    fun drawEnd() {
```

Comment:
Yes, this is tricky. All this was needed to avoid adding to the paths collection an eraser path that didn't actually erase anything... Do you get it?

Reply:
So it's an optimisation. I would argue that the cost of it it's higher than the gain. What do you think?

Reply:
I think it's actually needed, because of the consequences... If we didn't have this, we could erase parts of the picture without any paths (so, not erasing anything) and that would be "undoable", for instance... :-/

Reply:
I don't really see that as a problem, if the user erases an empty part of the mask, then so be it, they are on their own. Also, maybe by trying to be too clever, we might interfere with the user's flow, and confuse him/her. 
Last but not least, we would simplify this code, which is considerably complex already :D

Reply:
As agreed through call, we're keeping it as it is for now, and in any sign of poor performance, we will consider removing this treatment. Thanks!!